enhancements/ingress/optional-ingress.md

```diff
@@ -0,0 +1,225 @@
+- HyperShift creates the hosted cluster.
+- HyperShift skips the deployment of the ingress operator on the hosted control plane.
+- HyperShift creates the ingress operator's `ClusterVersion` without the ingress capability on the hosted cluster.
+- Hosted cluster's cluster version operator does not deploy the cluster ingress operator's payload (CRDs, RBACs, etc.).
+
+#### Setup AWS ALB as the new ingress
+
+- Cluster instance admin installs [AWS Load Balancer Operator](/enhancements/ingress/aws-load-balancer-operator.md).
+- Cluster instance admin sets up the default `AWSLoadBalancerController`.
+- Cluster instance user creates the `Ingress` resource for the default aws load balancer controller's ingress class.
+- AWS load balancer controller provisions an ALB on AWS.
+- Cluster instance admin can install [External DNS Operator](/enhancements/dns/externaldns-operator.md) to help cluster instance users set up convenient DNS names.
+
+### API Extensions
+
+N/A - no api extensions are being introduced in this proposal
+
+### Implementation Details/Notes/Constraints
+
+The name of the new capability is `Ingress`. It includes the cluster ingress operator and all its APIs listed below.
+The implementation of the new capability follows the instructions of [how to implement a new capability](/enhancements/installer/component-selection.md#how-to-implement-a-new-capability).
+
+#### HyperShift
+
+HyperShift doesn't pass all the ingress operator's payload to the cluster version operator.
+Certain manifests are filtered out, most importantly: the ingress operator's deployment.
+The ingress operator is considered as part of the hosted control plane, that is a resource not managed by the user hence deployed on a different cluster.
+Theoretically this may question the need for the cluster capability on HyperShift. However:
+- the cluster capability is a common interface for making OpenShift cluster components optional
+- the cluster capability is already implemented for some operators used by HyperShift (e.g. the console operator)
+- the capability provides a clear API which can be used by any party to understand whether the ingress is enabled
+- most importantly: [the ingress operator's APIs](#ingress-operator-apis) are still created by the cluster version operator
+
+#### Standalone OpenShift
+
+Making the ingress operator optional for the standalone OpenShift is a bigger scope than for HyperShift.
+The changes impact the components which are not used (or used differently) in HyperShift like the installer, the authentication and monitoring operators. To prevent the scope creep for [RFE-3395](https://issues.redhat.com/browse/RFE-3395), the installer should forbid the installation without the ingress capability in [the baseline or additional capabilities of the install config](https://github.com/openshift/installer/blob/master/pkg/types/installconfig.go#L479-L491).
```

Comment:
So if the user passes baselineCapabilitySet=None without also adding ingress to the additionalEnabledCapabilities, the installer would throw an error? That feels like an awkward usability story. What guarantees do we make about install-config portability across releases? Would this qualify as a breaking change?

Reply:
> So if the user passes baselineCapabilitySet=None without also adding ingress to the additionalEnabledCapabilities, the installer would throw an error? That feels like an awkward usability story.

Yes, that's quite misleading. The idea would be to not be able to specify the ingress capability in the ingress config at all. Make it sort of invisible until it's implemented for the standalone OpenShift.

> What guarantees do we make about install-config portability across releases? Would this qualify as a breaking change?

The upgrades seem safe to me. The ingress capability is always there no matter the baseline. A new release may unlock the disabling though. And this may be a problem for the downgrades. However I don't know well enough the downgrade logic. Will the install config be changed back to the snapshot of the X-1 version? If no, I don't quite understand how adding new capabilities work either.

Reply:
OK, so if the user sets the baseline to None the installer would also add ingress to the additionalEnabledCapabilities automatically?

Reply:
> > So if the user passes baselineCapabilitySet=None without also adding ingress to the additionalEnabledCapabilities, the installer would throw an error? That feels like an awkward usability story.
> 
> Yes, that's quite misleading. It would be ideal to be able to not specify the ingress capability in the ingress config at all. Make it sort of invisible until it's implemented for the standalone OpenShift.
> 
> > What guarantees do we make about install-config portability across releases? Would this qualify as a breaking change?
> 
> The upgrades seem safe to me. The ingress capability is always there no matter the baseline. A new release may unlock the disabling though. And this may be a problem for the downgrades. However I don't know well enough the downgrade logic.

The install config doesn't come into upgrades or downgrades. What I was worried about was a user who has built automation around an install config now getting different results. They won't get an error, so that's probably OK.

Reply:
> OK, so if the user sets the baseline to None the installer would also add ingress to the additionalEnabledCapabilities automatically?

Indeed, that can be an alternative to the throwing of an error. It would change the definition of `None` though.

Reply:
>  What I was worried about was a user who has built automation around an install config now getting different results. They won't get an error, so that's probably OK.

Such an automation, won't it be using the installer too? If no, who will be executing the install config?

Reply:
> OK, so if the user sets the baseline to None the installer would also add ingress to the additionalEnabledCapabilities automatically?

more likely might be to include the Ingress cap in the "None" capability set.  But i think regardless it's going to create some UX confusion, since ultimately the clusterversion.status.enabledCaps is going to show Ingress as an enabled cap (when spec.capSet==None and regardlss of whether spec.additionalCaps and status.additionalCaps have anything listed or not).

I imagine it will result in support cases of "i disabled all the caps but ingress is still reporting as enabled".  documentation/kcs will be needed.

Reply:
> This behavior seems to be similar to the implicitly enabled capability. The only difference is that here the capability is implicitly added at the installation time.

if we did it as i described, yeah. though i also realized we can't do it that way because then it would also be implicitly enabled on hypershift.

We need a way for it to be not enabled on hypershift, but force it to be enabled on non-hypershift.  Which i guess comes back to installer enforcement(either validation faliures, or always adding Ingress to the additionalEnabledCaps), unfortunately.

Reply:
> Which i guess comes back to installer enforcement(either validation failures, or always adding Ingress to the additionalEnabledCaps), unfortunately.
Yes. I'm not quite sure which of these 2 approaches is better though. The implicit adding may look a little smoother for the user.

Reply:
the implicit adding definitely seems smoother, i'm just not sure how it would function since it needs to discriminate between "traditional" and "hypershift" clusters.

Reply:
> What's the impact of disabling ingress on a standalone cluster? How much additional work would it be to make this change behave the same way for both types of clusters?

So far I found 3 components which can break on the standalone OpenShift only: the installer, the authentication operator and the motoring operator. I've listed them in `Implement the ingress capability for the standalone OpenShift` alternative section. Also, it's unclear how the absence of the ingress may impact the installation process: I didn't come up with a way to try this out.

Reply:
> I thought of the implementation in the openshift installer, not CVO. As far as I know the installer is not used by HyperShift. 

right, but there are two options we have discussed so far:

1) installer automatically adds "Ingress" as an `additionalEnabledCapability` in the installconfig (hypershift would not do this).  So that would be fine, it would work, it's just a bit awkward that the installer needs to do that/enforce that.  Once the cluster is installed, you can't disable a capability, so basically this would be fine unless we have other cluster creation workflow(besides hypershift) that do not go through the installer (I don't know what agent-installer and SNO do, for example?  but i think they still ultimately would go through this install path)

2) Add the Ingress cap to the "None" capset.  This would ensure ingress is always enabled even when all other caps are disabled(would show up as an implicitly enabled cap), but if we took this approach i'm not sure how we'd actually disable Ingress on hypershift.

so i'm just saying that i don't think my original suggestion (option 2) is actually viable, unless someone has some more thoughts around it.

Reply:
> the implicit adding definitely seems smoother, i'm just not sure how it would function since it needs to discriminate between "traditional" and "hypershift" clusters.

HyperShift's HostedControlPlane controller currently [manages a cluster-version operator Deployment directly][1], with no reliance on [how the CVO manages its own Deployment in stand-alone clusters][2].  So you could add an `--implicit-capabilities=Ingress` or whatever option to the CVO and only set that from the stand-alone manifest (until stand-alone implements the capability) without setting it from the HostedControlPlane controller.

[1]: https://github.com/openshift/hypershift/blob/203ffe95ed50591ccd059e1a6bbd638bd2014ab7/control-plane-operator/controllers/hostedcontrolplane/cvo/reconcile.go#L318-L336
[2]: https://github.com/openshift/cluster-version-operator/blob/f74d114b1b2843f045fdcdc63ca63fa4cf07da75/install/0000_00_cluster-version-operator_03_deployment.yaml

Reply:
>  So you could add an --required-capabilities=Ingress or whatever option to the CVO and only set that from the stand-alone manifest (until stand-alone implements the capability) without setting it from the HostedControlPlane controller.

My idea was: 
- do not touch CVO
- modify the installer's code so that the ingress capability is added [somewhere after the unmarshaling](https://github.com/openshift/installer/blob/master/pkg/types/defaults/installconfig.go#L32) from `install-config.yaml` (initially I thought of throwing an error at [the config validation](https://github.com/openshift/installer/blob/master/pkg/types/validation/installconfig.go#L59))
- let the installer pass the capabilities to CVO whatever the way it's currently done
- let the HyperShift control plane operator pass the capabilities verbatim to CVO whatever the way it's going to be designed

Reply:
Ben [raises concerns][1] about explicitly including `Ingress` in the `None` set potentially causing customer confusion, and floated implicitly enabling `Ingress` for stand-alone OpenShift  as a possible alternative.  I prefer that approach as well, and [gave one potential implementation][2].  But I'm not a decision-maker here, and I have no problem with you picking whichever as the primary direction, and then adding an alternatives section that lists the other options, the pros and cons of each, and why you decided you liked the one you picked best.

[1]: https://github.com/openshift/enhancements/pull/1415#discussion_r1218610977
[2]: https://github.com/openshift/enhancements/pull/1415#discussion_r1223620536

Reply:
I gave another thought to [Trevor's proposal](https://github.com/openshift/enhancements/pull/1415#discussion_r1223620536) and I think it makes sense because:
- it's a better fit for different OpenShift form factors: standalone, SNO, etc.
- the implicit capability enabling pairs well with this logic, so both are better to belong to the same repository
- it stays on the similar level of the UX as the change in the installer

I will update the EP with this proposal as the main one.